server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -249,6 +249,47 @@ public List<Pair<DataSegment, String>> retrieveUsedSegmentsAndCreatedDates(Strin
     );
   }
 
+  @Override
+  public List<DataSegment> retrieveUnusedSegmentsForExactIntervalAndVersion(
+      String dataSource,
+      Interval interval,
+      String version
+  )
+  {
+    StringBuilder sb = new StringBuilder();
+    sb.append(StringUtils.format("SELECT payload FROM %s", dbTables.getSegmentsTable()));
+    sb.append(" WHERE used = :used AND dataSource = :dataSource AND version = :version");
+    sb.append(StringUtils.format(" AND start = :start AND %1$send%1$s = :end", connector.getQuoteString()));
```

Comment:
Don't use this pattern as it makes the query difficult to read. Do this instead:

```
final String sql = "the whole sql";
handle.createQuery(StringUtils.format(sql, all the args))
```

Reply:
Done